src/NUnitFramework/framework/Constraints/DictionaryContainsKeyValuePairConstraint.cs

```diff
@@ -70,6 +71,33 @@ private bool Matches(object? actual)
                 return false;
             }
 
+            // If 'actual' implements IEnumerable<KeyValuePair<TKey, TValue>>, we can find the value using that interface
+            var enumerableType = actual.GetType().GetInterface("IEnumerable`1");
+            if (enumerableType is not null)
+            {
+                var enumerableInnerType = enumerableType.GenericTypeArguments[0];
+                if (enumerableInnerType.IsGenericType && enumerableInnerType.GetGenericTypeDefinition() == typeof(KeyValuePair<,>))
+                {
+                    var keyGetter = enumerableInnerType.GetProperty("Key");
+                    var valueGetter = enumerableInnerType.GetProperty("Value");
+
+                    Debug.Assert(keyGetter is not null && valueGetter is not null);
+
+                    foreach (object item in (IEnumerable)actual)
+                    {
+                        var expectedKvp = new KeyValuePair<object, object?>(_expected.Key, _expected.Value);
+                        var actualKvp = new KeyValuePair<object, object?>(keyGetter.GetValue(item)!, valueGetter.GetValue(item));
+
+                        if (ItemsEqual(actualKvp, expectedKvp))
+                        {
+                            return true;
+                        }
```

Comment:
Is any of this needed? The code in the existing code above, even though it checks for `IDictionary` could as well only check for `IEnumerable<KeyValuePair>` as that is all that it uses.

Reply:
That's a good point. I changed the entire constraint to remove any dictionary-ness and instead make it work with any "key-value pair provider".